charms/kfp-persistence/src/relation_components.py

```diff
@@ -0,0 +1,127 @@
+
+    def get_data(self) -> Union[List[dict], dict]:
+        """Validates and returns the data stored in this relation.
+        Validation asserts that there is data for exactly one related app.
+        Raises ErrorWithStatus if data can not be returned.
+        """
+        try:
+            interface = self.get_interface()
+        # TODO: These messages should be tested and cleaned up
+        except (NoVersionsListed, UnversionedRelation) as err:
+            raise ErrorWithStatus(str(err), WaitingStatus) from err
+        except NoCompatibleVersions as err:
+            raise ErrorWithStatus(str(err), BlockedStatus) from err
+        except Exception as err:
+            raise ErrorWithStatus(f"Caught unknown error: '{str(err)}'", BlockedStatus) from err
+
+        if interface is None:
+            msg = f"Missing required data from 1 application on relation {self._relation_name}"
+            raise ErrorWithStatus(str(msg), BlockedStatus)
+
+        try:
+            # TODO: This might be multiple values (one for each related app).  This is returning a
+            #  list
+            unpacked_data = list(interface.get_data().values())
+        except ValidationError as val_error:
+            # Validation in .get_data() ensures if data is populated, it matches the schema and is
+            # not incomplete
+            msg = (
+                "Got ValidationError when interpreting data on relation"
+                f"{self._relation_name}: {val_error}"
+            )
+            raise ErrorWithStatus(msg, BlockedStatus) from val_error
+
+        self.validate_data(unpacked_data)
+
+        # If relation supports exactly 1 relation, return just that relation's data.  Else, return
+        # as a list.
+        if self._minimum_related_applications == self._maximum_related_applications:
+            return unpacked_data[0]
+        else:
+            return unpacked_data
+
+    def get_interface(self) -> Optional[SerializedDataInterface]:
+        """Returns the SerializedDataInterface object for this interface."""
+        return get_interface(self._charm, self._relation_name)
+
+    def get_status(self) -> StatusBase:
+        """Returns the status of this relation.
+        Use this in the charm to inspect the state of the relation and its data.
+        Will return:
+            * BlockedStatus: if we have no compatible versions on the relation, or no related
+                             app
+            * WaitingStatus: if we have not yet received a version from the opposite relation
+            * ActiveStatus: if:
+                * nothing is related to us (as there is no work to do)
+                * we have one or more relations, and we have sent data to all of them
+        """
+        try:
+            # If we successfully get data
+            self.get_data()
+        except ErrorWithStatus as err:
+            return err.status
+
+        return ActiveStatus()
+
+    def validate_data(self, data: List[dict]):
```

Comment:
As @ca-scribner mentioned in our of discussions, is `validate_data` a proper name for this? My understanding is that this is validating the correct number of relations and not relation data

Reply:
Also, just re-using component here. I suspect this function is not complete. Currently, it checks max number of relations, but it can also check contents of relation.